pkg/runtime-manager/server/cri/criserver.go

```diff
@@ -0,0 +1,129 @@
+)
+
+type RuntimeManagerCriServer struct {
+	dispatcher    *dispatcher.RuntimeHookDispatcher
+	runtimeClient runtimeapi.RuntimeServiceClient
+}
+
+func NewRuntimeManagerCriServer(dispatcher *dispatcher.RuntimeHookDispatcher) *RuntimeManagerCriServer {
+	criInterceptor := &RuntimeManagerCriServer{
+		dispatcher: dispatcher,
+	}
+	return criInterceptor
+}
+
+func (ci *RuntimeManagerCriServer) Name() string {
+	return "RuntimeManagerCriServer"
+}
+
+func (ci *RuntimeManagerCriServer) Setup() error {
+	if err := ci.initBackendServer(utils.DefaultContainerdSocketPath); err != nil {
+		return err
+	}
+	lis, err := net.Listen("unix", utils.DefaultRuntimeManagerSocketPath)
+	if err != nil {
+		klog.Errorf("fail to create the lis %v", err)
+		return err
+	}
+	grpcServer := grpc.NewServer()
+	runtimeapi.RegisterRuntimeServiceServer(grpcServer, ci)
+	err = grpcServer.Serve(lis)
+	return err
+}
+
+func (ci *RuntimeManagerCriServer) getRuntimeHookInfo(serviceType RuntimeServiceType) (config.RuntimeRequestPath,
+	resource_executor.RuntimeResourceType) {
+	switch serviceType {
+	case RunPodSandbox:
+		return config.RunPodSandbox, resource_executor.RuntimePodResource
+	case CreateContainer:
+		// No Nook point in create container, but we need store the container info during container create
+		return config.NoneRuntimeHookPath, resource_executor.RuntimeContainerResource
+	case StartContainer:
+		return config.StartContainer, resource_executor.RuntimeContainerResource
+	case StopContainer:
+		return config.StopContainer, resource_executor.RuntimeContainerResource
+	case UpdateContainerResources:
+		return config.UpdateContainerResources, resource_executor.RuntimeContainerResource
+	}
+	return config.NoneRuntimeHookPath, resource_executor.RuntimeNoopResource
+}
+
+func (ci *RuntimeManagerCriServer) interceptRuntimeRequest(serviceType RuntimeServiceType,
+	ctx context.Context, request interface{}, handler grpc.UnaryHandler) (interface{}, error) {
+
+	runtimeHookPath, runtimeResourceType := ci.getRuntimeHookInfo(serviceType)
+	resourceExecutor := resource_executor.NewRuntimeResourceExecutor(runtimeResourceType)
+
+	if err := resourceExecutor.ParseRequest(request); err != nil {
+		klog.Errorf("fail to parse request %v %v", request, err)
+	}
+
+	// pre call hook server
+	if _, err := ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, resourceExecutor.GenerateHookRequest()); err != nil {
+		klog.Errorf("fail to call hook server %v", err)
+	}
+
+	// call the backend runtime engine
+	res, err := handler(ctx, request)
+	if err == nil {
+		klog.Infof("%v %v success", res, err)
+		// store checkpoint info basing request
+		// checkpoint only when response success
+		if err := resourceExecutor.ResourceCheckPoint(res); err != nil {
+			klog.Errorf("fail to checkpoint %v", err)
+		}
+	} else {
+		klog.Errorf("%v %v", res, err)
+	}
+
+	// post call hook server
+	ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, resourceExecutor.GenerateHookRequest())
```

Comment:
Here is a detail, if the above handler fails, the post hook needs to be aware of what the result of the handler is.

Reply:
> Here is a detail, if the above handler fails, the post hook needs to be aware of what the result of the handler is.

yup,  there is  a TODO in protocol to hookserver:

message ContainerResourceHookRequest {
  ...
  LinuxContainerResources pod_resources = 5;
  // TODO: add the error info from containerd/dockerd
}

See 
https://github.com/koordinator-sh/koordinator/blob/main/apis/runtime/v1alpha1/api.proto#L127